### PR TITLE
refactor: safeset for controller cache resource.

### PR DESCRIPTION
**Pull request by qicz**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Following #4454, optimize the resources cached with the safety set.

Release Notes: Yes/No

### CODE REVIEW

**Comment by arkodg** [comment]
@qicz can you elaborate the issue in the description

File: internal/provider/kubernetes/resource.go [hunk 1]
```diff
@@ -64,28 +66,60 @@ type resourceMappings struct {
 	httpRouteFilters map[utils.NamespacedNameWithGroupKind]*egv1a1.HTTPRouteFilter
 }
 
+type safeSet[T comparable] struct {
```
Comment (zirain) [inline internal/provider/kubernetes/resource.go:69]: this should live in utils package with some  tests

**Comment by qicz** [comment]
> @qicz can you elaborate the issue in the description

done

File: internal/utils/safeset/safeset.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type SafeSet[T comparable] struct {
+	lock   sync.RWMutex
+	Values sets.Set[T]
+}
+
+func NewSafeSet[T comparable](items ...T) *SafeSet[T] {
```
Comment (zirain) [inline internal/utils/safeset/safeset.go:19]:
```suggestion
func New[T comparable](items ...T) *SafeSet[T] {
```

File: internal/utils/safeset/safeset.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type SafeSet[T comparable] struct {
+	lock   sync.RWMutex
+	Values sets.Set[T]
```
Comment (zirain) [inline internal/utils/safeset/safeset.go:16]: why expose?

**Comment by zirain** [comment]
@qicz can you share more details about why we need a lock here?

**Pull request closed without merging**